From afef331e1a2be2ef2f4f25ee9dfc711aadb5f2b8 Mon Sep 17 00:00:00 2001
From: Robin Reckmann <robin.reckmann@gmail.com>
Date: Thu, 26 Mar 2026 00:24:06 +0900
Subject: drm/panel: add dwo do0200fat07 panel driver

Add support for the dwo do0200fat07 240x536 panel found on the comma mici.
---
 drivers/gpu/drm/panel/Kconfig                 |  10 +
 drivers/gpu/drm/panel/Makefile                |   1 +
 drivers/gpu/drm/panel/panel-dwo-do0200fat07.c | 333 ++++++++++++++++++
 3 files changed, 344 insertions(+)
 create mode 100644 drivers/gpu/drm/panel/panel-dwo-do0200fat07.c

diff --git a/drivers/gpu/drm/panel/Kconfig b/drivers/gpu/drm/panel/Kconfig
--- a/drivers/gpu/drm/panel/Kconfig
+++ b/drivers/gpu/drm/panel/Kconfig
@@ -136,6 +136,16 @@
 	help
 	  Say Y here if you want to enable support for ChipWealth
 	  CH13726A-based display panels.
+
+config DRM_PANEL_DWO_DO0200FAT07
+   tristate "DWO DO0200FAT07 panel"
+   depends on OF
+   depends on DRM_MIPI_DSI
+   depends on BACKLIGHT_CLASS_DEVICE
+   select VIDEOMODE_HELPERS
+   help
+     Say Y here if you want to enable support for the DWo DO0200FAT07
+     panel.
 
 config DRM_PANEL_EBBG_FT8719
 	tristate "EBBG FT8719 panel driver"
diff --git a/drivers/gpu/drm/panel/Makefile b/drivers/gpu/drm/panel/Makefile
--- a/drivers/gpu/drm/panel/Makefile
+++ b/drivers/gpu/drm/panel/Makefile
@@ -16,6 +16,7 @@
 obj-$(CONFIG_DRM_PANEL_LVDS) += panel-lvds.o
 obj-$(CONFIG_DRM_PANEL_SIMPLE) += panel-simple.o
 obj-$(CONFIG_DRM_PANEL_EDP) += panel-edp.o
+obj-$(CONFIG_DRM_PANEL_DWO_DO0200FAT07) += panel-dwo-do0200fat07.o
 obj-$(CONFIG_DRM_PANEL_EBBG_FT8719) += panel-ebbg-ft8719.o
 obj-$(CONFIG_DRM_PANEL_ELIDA_KD35T133) += panel-elida-kd35t133.o
 obj-$(CONFIG_DRM_PANEL_FEIXIN_K101_IM2BA02) += panel-feixin-k101-im2ba02.o
diff --git a/drivers/gpu/drm/panel/panel-dwo-do0200fat07.c b/drivers/gpu/drm/panel/panel-dwo-do0200fat07.c
new file mode 100644
--- /dev/null
+++ b/drivers/gpu/drm/panel/panel-dwo-do0200fat07.c
@@ -0,0 +1,334 @@
+// SPDX-License-Identifier: GPL-2.0-only
+// Copyright (c) 2026 Robin Reckmann <robin.reckmann@gmail.com>
+
+#include <linux/backlight.h>
+#include <linux/delay.h>
+#include <linux/gpio/consumer.h>
+#include <linux/mod_devicetable.h>
+#include <linux/module.h>
+#include <linux/regulator/consumer.h>
+
+#include <video/mipi_display.h>
+
+#include <drm/drm_mipi_dsi.h>
+#include <drm/drm_modes.h>
+#include <drm/drm_of.h>
+#include <drm/drm_panel.h>
+#include <drm/drm_probe_helper.h>
+
+struct dwo_do0200fat07 {
+	struct drm_panel panel;
+	struct mipi_dsi_device *dsi;
+	struct gpio_desc *reset_gpio;
+	struct regulator *vddio;
+	struct regulator *vdd;
+	enum drm_panel_orientation orientation;
+};
+
+static inline struct dwo_do0200fat07 *
+to_dwo_do0200fat07(struct drm_panel *panel)
+{
+	return container_of(panel, struct dwo_do0200fat07, panel);
+}
+
+static void dwo_do0200fat07_reset(struct dwo_do0200fat07 *ctx)
+{
+	gpiod_set_value_cansleep(ctx->reset_gpio, 0);
+	usleep_range(1000, 2000);
+	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+	usleep_range(1000, 2000);
+	gpiod_set_value_cansleep(ctx->reset_gpio, 0);
+	usleep_range(10000, 11000);
+}
+
+static int dwo_do0200fat07_power_on(struct dwo_do0200fat07 *ctx)
+{
+	int ret;
+
+	ret = regulator_enable(ctx->vddio);
+	if (ret)
+		return ret;
+	usleep_range(3000, 5000);
+
+	ret = regulator_enable(ctx->vdd);
+	if (ret) {
+		regulator_disable(ctx->vddio);
+		return ret;
+	}
+	usleep_range(3000, 5000);
+	return 0;
+}
+
+static void dwo_do0200fat07_power_off(struct dwo_do0200fat07 *ctx)
+{
+	regulator_disable(ctx->vdd);
+	regulator_disable(ctx->vddio);
+}
+
+static int dwo_do0200fat07_init_sequence(struct dwo_do0200fat07 *ctx)
+{
+	struct mipi_dsi_multi_context dsi_ctx = { .dsi = ctx->dsi };
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xfe, 0x05);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x91, 0x00);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x00, 0x00);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xfe, 0x01);
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0x1c, 0x99);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xfe, 0x00);
+
+	mipi_dsi_dcs_set_pixel_format_multi(
+		&dsi_ctx,
+		MIPI_DCS_PIXEL_FMT_24BIT | (MIPI_DCS_PIXEL_FMT_24BIT << 4));
+	mipi_dsi_dcs_set_display_brightness_multi(&dsi_ctx, 0x0000);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, MIPI_DCS_SET_ADDRESS_MODE, 0x00);
+
+	mipi_dsi_dcs_set_tear_on_multi(&dsi_ctx,
+				       MIPI_DSI_DCS_TEAR_MODE_VHBLANK);
+
+	mipi_dsi_dcs_write_seq_multi(&dsi_ctx, 0xc2, 0x03);
+
+	mipi_dsi_dcs_exit_sleep_mode_multi(&dsi_ctx);
+	mipi_dsi_msleep(&dsi_ctx, 100);
+
+	return dsi_ctx.accum_err;
+}
+
+static int dwo_do0200fat07_prepare(struct drm_panel *panel)
+{
+	struct dwo_do0200fat07 *ctx = to_dwo_do0200fat07(panel);
+	struct device *dev = &ctx->dsi->dev;
+	int ret;
+
+	ret = dwo_do0200fat07_power_on(ctx);
+	if (ret) {
+		dev_err(dev, "Failed to power on: %d\n", ret);
+		return ret;
+	}
+
+	dwo_do0200fat07_reset(ctx);
+
+	ret = dwo_do0200fat07_init_sequence(ctx);
+	if (ret < 0) {
+		dev_err(dev, "Failed to initialize panel: %d\n", ret);
+		gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+		dwo_do0200fat07_power_off(ctx);
+		return ret;
+	}
+
+	return 0;
+}
+
+static int dwo_do0200fat07_unprepare(struct drm_panel *panel)
+{
+	struct dwo_do0200fat07 *ctx = to_dwo_do0200fat07(panel);
+
+	gpiod_set_value_cansleep(ctx->reset_gpio, 1);
+	usleep_range(1000, 2000);
+	dwo_do0200fat07_power_off(ctx);
+
+	return 0;
+}
+
+static int dwo_do0200fat07_enable(struct drm_panel *panel)
+{
+	struct dwo_do0200fat07 *ctx = to_dwo_do0200fat07(panel);
+	struct mipi_dsi_multi_context c = { .dsi = ctx->dsi };
+
+	mipi_dsi_dcs_set_display_on_multi(&c);
+	mipi_dsi_msleep(&c, 20);
+
+	return c.accum_err;
+}
+
+static int dwo_do0200fat07_disable(struct drm_panel *panel)
+{
+	struct dwo_do0200fat07 *ctx = to_dwo_do0200fat07(panel);
+	struct mipi_dsi_multi_context c = { .dsi = ctx->dsi };
+
+	mipi_dsi_dcs_set_display_off_multi(&c);
+	mipi_dsi_msleep(&c, 20);
+
+	mipi_dsi_dcs_enter_sleep_mode_multi(&c);
+	mipi_dsi_msleep(&c, 120);
+
+	return c.accum_err;
+}
+
+static const struct drm_display_mode dwo_do0200fat07_mode = {
+	.clock = (240 + 20 + 20 + 40) * (536 + 20 + 4 + 12) * 60 / 1000,
+	.hdisplay = 240,
+	.hsync_start = 240 + 20,
+	.hsync_end = 240 + 20 + 20,
+	.htotal = 240 + 20 + 20 + 40,
+	.vdisplay = 536,
+	.vsync_start = 536 + 20,
+	.vsync_end = 536 + 20 + 4,
+	.vtotal = 536 + 20 + 4 + 12,
+	.width_mm = 20,
+	.height_mm = 44,
+	.type = DRM_MODE_TYPE_DRIVER,
+};
+
+static int dwo_do0200fat07_get_modes(struct drm_panel *panel,
+				     struct drm_connector *connector)
+{
+	struct drm_display_mode *mode;
+
+	mode = drm_mode_duplicate(connector->dev, &dwo_do0200fat07_mode);
+	if (!mode)
+		return -ENOMEM;
+
+	drm_mode_set_name(mode);
+	mode->type = DRM_MODE_TYPE_DRIVER | DRM_MODE_TYPE_PREFERRED;
+	drm_mode_probed_add(connector, mode);
+
+	connector->display_info.bpc = 8;
+	connector->display_info.width_mm = dwo_do0200fat07_mode.width_mm;
+	connector->display_info.height_mm = dwo_do0200fat07_mode.height_mm;
+
+	return 1;
+}
+
+static enum drm_panel_orientation
+dwo_do0200fat07_get_orientation(struct drm_panel *panel)
+{
+	struct dwo_do0200fat07 *ctx = to_dwo_do0200fat07(panel);
+
+	return ctx->orientation;
+}
+
+static const struct drm_panel_funcs dwo_do0200fat07_panel_funcs = {
+	.prepare = dwo_do0200fat07_prepare,
+	.unprepare = dwo_do0200fat07_unprepare,
+	.enable = dwo_do0200fat07_enable,
+	.disable = dwo_do0200fat07_disable,
+	.get_modes = dwo_do0200fat07_get_modes,
+	.get_orientation = dwo_do0200fat07_get_orientation,
+};
+
+static int dwo_do0200fat07_bl_update_status(struct backlight_device *bl)
+{
+	struct mipi_dsi_device *dsi = bl_get_data(bl);
+	u16 brightness = backlight_get_brightness(bl);
+	int ret;
+
+	dsi->mode_flags &= ~MIPI_DSI_MODE_LPM;
+
+	ret = mipi_dsi_dcs_set_display_brightness(dsi, brightness);
+
+	dsi->mode_flags |= MIPI_DSI_MODE_LPM;
+
+	return ret;
+}
+
+static const struct backlight_ops dwo_do0200fat07_bl_ops = {
+	.update_status = dwo_do0200fat07_bl_update_status,
+};
+
+static struct backlight_device *
+dwo_do0200fat07_create_backlight(struct mipi_dsi_device *dsi)
+{
+	struct device *dev = &dsi->dev;
+	const struct backlight_properties props = {
+		.type = BACKLIGHT_RAW,
+		.brightness = 255,
+		.max_brightness = 255,
+	};
+
+	return devm_backlight_device_register(dev, dev_name(dev), dev, dsi,
+					      &dwo_do0200fat07_bl_ops, &props);
+}
+
+static int dwo_do0200fat07_probe(struct mipi_dsi_device *dsi)
+{
+	struct device *dev = &dsi->dev;
+	struct dwo_do0200fat07 *ctx;
+	int ret;
+
+	ctx = devm_drm_panel_alloc(dev, struct dwo_do0200fat07, panel,
+				   &dwo_do0200fat07_panel_funcs,
+				   DRM_MODE_CONNECTOR_DSI);
+	if (IS_ERR(ctx))
+		return PTR_ERR(ctx);
+
+	ctx->vddio = devm_regulator_get(dev, "vddio");
+	if (IS_ERR(ctx->vddio))
+		return dev_err_probe(dev, PTR_ERR(ctx->vddio),
+				     "Failed to get vddio-supply\n");
+
+	ctx->vdd = devm_regulator_get(dev, "vdd");
+	if (IS_ERR(ctx->vdd))
+		return dev_err_probe(dev, PTR_ERR(ctx->vdd),
+				     "Failed to get vdd-supply\n");
+
+	ctx->reset_gpio = devm_gpiod_get(dev, "reset", GPIOD_OUT_HIGH);
+	if (IS_ERR(ctx->reset_gpio))
+		return dev_err_probe(dev, PTR_ERR(ctx->reset_gpio),
+				     "Failed to get reset-gpios\n");
+
+	ctx->dsi = dsi;
+	mipi_dsi_set_drvdata(dsi, ctx);
+
+	dsi->lanes = 1;
+	dsi->format = MIPI_DSI_FMT_RGB888;
+	dsi->mode_flags = MIPI_DSI_MODE_VIDEO | MIPI_DSI_MODE_VIDEO_BURST |
+			  MIPI_DSI_CLOCK_NON_CONTINUOUS | MIPI_DSI_MODE_LPM;
+
+	ctx->panel.prepare_prev_first = true;
+
+	ctx->panel.backlight = dwo_do0200fat07_create_backlight(dsi);
+	if (IS_ERR(ctx->panel.backlight))
+		return dev_err_probe(dev, PTR_ERR(ctx->panel.backlight),
+				     "Failed to create backlight\n");
+
+	ret = drm_of_get_panel_orientation(dev->of_node, &ctx->orientation);
+	if (ret) {
+		return dev_err_probe(dev, ret, "Failed to get orientation\n");
+	}
+
+	drm_panel_add(&ctx->panel);
+
+	ret = mipi_dsi_attach(dsi);
+	if (ret < 0) {
+		drm_panel_remove(&ctx->panel);
+		return dev_err_probe(dev, ret,
+				     "Failed to attach to DSI host\n");
+	}
+
+	return 0;
+}
+
+static void dwo_do0200fat07_remove(struct mipi_dsi_device *dsi)
+{
+	struct dwo_do0200fat07 *ctx = mipi_dsi_get_drvdata(dsi);
+	int ret;
+
+	ret = mipi_dsi_detach(dsi);
+	if (ret < 0)
+		dev_err(&dsi->dev, "Failed to detach from DSI host: %d\n", ret);
+
+	drm_panel_remove(&ctx->panel);
+}
+
+static const struct of_device_id dwo_do0200fat07_of_match[] = {
+	{ .compatible = "dwo,do0200fat07" },
+	{ /* sentinel */ }
+};
+MODULE_DEVICE_TABLE(of, dwo_do0200fat07_of_match);
+
+static struct mipi_dsi_driver dwo_do0200fat07_driver = {
+	.probe = dwo_do0200fat07_probe,
+	.remove = dwo_do0200fat07_remove,
+	.driver = {
+		.name = "panel-dwo-do0200fat07",
+		.of_match_table = dwo_do0200fat07_of_match,
+	},
+};
+module_mipi_dsi_driver(dwo_do0200fat07_driver);
+
+MODULE_AUTHOR("Robin Reckmann <robin.reckmann@gmail.com>");
+MODULE_DESCRIPTION("DRM driver for DWo DO0200FAT07 panel");
+MODULE_LICENSE("GPL");
